File: Testing/tests/client/Tests/Linq/TestLinqOrderingOperators.cs
using Bridge.QUnit;
using ClientTestLibrary.Utilities;
using System.Linq;

namespace ClientTestLibrary.Linq
{
    internal class TestLinqOrderingOperators
    {
        public static void Test(Assert assert)
        {
            assert.Expect(8);

            // TEST
            var words = new string[] { "ab2", "ac", "a", "ab12", "", "ab", "bac", "z" };
            var sortedWords = (from word in words
                               orderby word
                               select word).ToArray();
            assert.DeepEqual(sortedWords, new[] { "", "a", "ab", "ab12", "ab2", "ac", "bac", "z" }, "Order by words");

            // TEST
            var sortedWordsByLength = (from word in words
                                       orderby word.Length
                                       select word).ToArray();
            assert.DeepEqual(sortedWordsByLength, new[] { "", "a", "z", "ac", "ab", "ab2", "bac", "ab12" }, "Order by word length");

            // TEST
            var sortedPersonsByName = (from p in Person.GetPersons()
                                       orderby p.Name
                                       select p.Name).ToArray();
            assert.DeepEqual(sortedPersonsByName, new[] { "Billy", "Dora", "Frank", "Ian", "John", "Mary", "Nemo", "Zeppa" }, "Order by person names");

            // TODO test with System.StringComparison

            // TEST
            var doubles = new double[] { 1.0, -0.7, 2.1, 0.9, 1.4, 2.9 };
            var sortedDoubles = (from d in doubles
                                 orderby d descending
                                 select d).ToArray();
            assert.DeepEqual(sortedDoubles, new[] { 2.9, 2.1, 1.4, 1.0, 0.9, -0.7 }, "Order by descending double");

            // TEST
            var sortedPersonsByCountDesc = (from p in Person.GetPersons()
                                            orderby p.Count descending
                                            select p.Count).ToArray();
            assert.DeepEqual(sortedPersonsByCountDesc, new[] { 3000, 700, 700, 550, 500, 300, 100, 50 }, "Order by person count descending");

            // TEST
            var sortedWordsByLengthAndLetters = (from word in words
                                                 orderby word.Length, word
                                                 select word).ToArray();
            assert.DeepEqual(sortedWordsByLengthAndLetters, new[] { "", "a", "z", "ab", "ac", "ab2", "bac", "ab12" }, "Order by word length then by letters");

            // TEST
            var sortedWordsByLengthAndLettersLambda = words.OrderBy(x => x.Length).ThenByDescending(x => x).ToArray();
            assert.DeepEqual(sortedWordsByLengthAndLettersLambda, new[] { "", "z", "a", "ac", "ab", "bac", "ab2", "ab12" }, "Order by word length then by letters as lambda");

            // TEST
            // var numbers = new[] { 2, 4, 6, 1, 5, 7, 9, 0, 8, 3};
            var numbers = new[] { 2, 4, 6, 1, 5 };
            var numbersReversed = numbers.Reverse<int>().ToArray();
            assert.DeepEqual(numbersReversed, new[] { 5, 1, 6, 4, 2 }, "Reverse() numbers");
        }
    }
}

File: Testing/www/js/testLinqOrderingOperators.js
/* global Bridge */

"use strict";

Bridge.define('ClientTestLibrary.Linq.TestLinqOrderingOperators', {
    statics: {
        test: function (assert) {
            assert.expect(8);

            // TEST
            var words = ["ab2", "ac", "a", "ab12", "", "ab", "bac", "z"];
            var sortedWords = (Bridge.Linq.Enumerable.from(words).orderBy(function (word) {
                return word;
            })).toArray();
            assert.deepEqual(sortedWords, ["", "a", "ab", "ab12", "ab2", "ac", "bac", "z"], "Order by words");

            // TEST
            var sortedWordsByLength = (Bridge.Linq.Enumerable.from(words).orderBy(function (word) {
                return word.length;
            })).toArray();
            assert.deepEqual(sortedWordsByLength, ["", "a", "z", "ac", "ab", "ab2", "bac", "ab12"], "Order by word length");

            // TEST
            var sortedPersonsByName = (Bridge.Linq.Enumerable.from(ClientTestLibrary.Utilities.Person.getPersons()).orderBy(function (p) {
                return p.getName();
            }).select(function (p) {
                return p.getName();
            })).toArray();
            assert.deepEqual(sortedPersonsByName, ["Billy", "Dora", "Frank", "Ian", "John", "Mary", "Nemo", "Zeppa"], "Order by person names");

            // TODO test with System.StringComparison

            // TEST
            var doubles = [1.0, -0.7, 2.1, 0.9, 1.4, 2.9];
            var sortedDoubles = (Bridge.Linq.Enumerable.from(doubles).orderByDescending(function (d) {
                return d;
            })).toArray();
            assert.deepEqual(sortedDoubles, [2.9, 2.1, 1.4, 1.0, 0.9, -0.7], "Order by descending double");

            // TEST
            var sortedPersonsByCountDesc = (Bridge.Linq.Enumerable.from(ClientTestLibrary.Utilities.Person.getPersons()).orderByDescending(function (p) {
                return p.getCount();
            }).select(function (p) {
                return p.getCount();
            })).toArray();
            assert.deepEqual(sortedPersonsByCountDesc, [3000, 700, 700, 550, 500, 300, 100, 50], "Order by person count descending");

            // TEST
            var sortedWordsByLengthAndLetters = (Bridge.Linq.Enumerable.from(words).orderBy(function (word) {
                return word.length;
            }).thenBy(function (word) {
                return word;
            })).toArray();
            assert.deepEqual(sortedWordsByLengthAndLetters, ["", "a", "z", "ab", "ac", "ab2", "bac", "ab12"], "Order by word length then by letters");

            // TEST
            var sortedWordsByLengthAndLettersLambda = Bridge.Linq.Enumerable.from(words).orderBy(function (x) {
                return x.length;
            }).thenByDescending(function (x) {
                return x;
            }).toArray();
            assert.deepEqual(sortedWordsByLengthAndLettersLambda, ["", "z", "a", "ac", "ab", "bac", "ab2", "ab12"], "Order by word length then by letters as lambda");

            // TEST
            // var numbers = new[] { 2, 4, 6, 1, 5, 7, 9, 0, 8, 3};
            var numbers = [2, 4, 6, 1, 5];
            var numbersReversed = Bridge.Linq.Enumerable.from(numbers).reverse().toArray();
            assert.deepEqual(numbersReversed, [5, 1, 6, 4, 2], "Reverse() numbers");
        }
    }
});



Bridge.init();